### PR TITLE
Introduced interface for IOconfFile #699

### PR DESCRIPTION
**Pull request by LarsWithCA**
Injecting this into all classes/methods that need the configuration. 
SerialNumberMapper and MCUBoard accept a null configuration.

### CODE REVIEW

File: CA_DataUploader/Program.cs [hunk 1]
```diff
@@ -18,10 +19,11 @@ static async Task MainAsync(string[] args)
         {
             try
             {
-                CALog.LogInfoAndConsoleLn(LogID.A, RpiVersion.GetWelcomeMessage($"Upload temperature data to cloud"));
+                var loglevel = IOconfFileLoader.FileExists() ? IOconfFile.Instance.GetOutputLevel() : IOconfLoopName.Default.LogLevel;
+                CALog.LogInfoAndConsoleLn(LogID.A, RpiVersion.GetWelcomeMessage($"Upload temperature data to cloud", loglevel));
                 Console.WriteLine("Initializing...");
                 Redundancy.RegisterSystemExtensions();
-                var serial = new SerialNumberMapper();
+                var serial = new SerialNumberMapper(null);
```
Comment (freddyrios) [inline CA_DataUploader/Program.cs:26]: I think not passing a loaded ioconf would give trouble if answering yes to "Do you want to skip IO.conf setup? (yes, no)" in IOconfSetup.UpdateIOconf (answering no causes config to be updated and boards to be mapped again during reload).

**Review by freddyrios — APPROVED** [review]

File: CA_DataUploaderLib/IOconf/IOconfFile.cs [hunk 2]
```diff
@@ -59,33 +62,34 @@ public static ConnectionInfo GetConnectionInfo()
...
+        public IEnumerable<IOconfOven> GetOven() => GetEntries<IOconfOven>();
+        public IEnumerable<IOconfAlert> GetAlerts() => GetEntries<IOconfAlert>();
+        public IEnumerable<IOconfMath> GetMath() => GetEntries<IOconfMath>();
+        public IEnumerable<IOconfFilter> GetFilters() => GetEntries<IOconfFilter>();
+        public IEnumerable<IOconfOutput> GetOutputs() => GetEntries<IOconfOutput>();
+        public IEnumerable<IOconfState> GetStates() => GetEntries<IOconfState>();
+        public IEnumerable<IOconfInput> GetInputs() => GetEntries<IOconfInput>();
+        public IEnumerable<T> GetEntries<T>() => Table.OfType<T>();
+        public string GetRawFile() => string.Join(Environment.NewLine, RawLines);
+        ///<remr filters and math it returs the board state of all their sources.</remarks>
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply CA_DataUploaderLib/IOconf/IOconfFile.cs:86]: fixed

File: CA_DataUploaderLib/IOconf/IOconfFile.cs [hunk 3]
```diff
@@ -6,31 +6,34 @@
...
             {
                 Reload();
             }
         }
 
-        public static void Reload()
-        { 
+        public void Reload()
+        {
             // the separate IOconfFileLoader can be used by callers to expand the IOconfFile before the IOconfFile initialization / static ctor rejects the custom entries.
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply CA_DataUploaderLib/IOconf/IOconfFile.cs:27]: fixed

File: CA_DataUploaderLib/SwitchBoardController.cs [hunk 4]
```diff
@@ -9,25 +9,24 @@ public sealed class SwitchBoardController : BaseSensorBox
...
             //we ignore remote boards and boards missing during the start sequence (as we don't have auto reconnect logic yet for those). Note the BaseSensorBox already reports the missing local boards.
-            foreach (var port in ports.Where(p => p.Map.IsLocalBoard && p.Map.McuBoard != null))
+            foreach (var port in ioconf.GetEntries<IOconfOut230Vac>().Where(p => p.Map.IsLocalBoard && p.Map.McuBoard != null))
                 RegisterBoardWriteActions(port.Map.McuBoard!, port, 0, port.Name + "_onoff", GetCommand);
 
             static string GetCommand(int portNumber, double target) => target > 0.0 ? $"p{portNumber} on 3 {target:0%}" : $"p{portNumber} off";
         }
 
-        public static void Initialize(CommandHandler cmd)
+        public static void Initialize(IIOconf ioconf, CommandHandler cmd)
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply CA_DataUploaderLib/SwitchBoardController.cs:25]: note to self: regarding validation

File: CA_DataUploader/Program.cs (hunk 1 above)
Comment (LarsWithCA) [reply CA_DataUploader/Program.cs:26]: fixed

File: CA_DataUploaderLib/IOconf/IOconfTemp.cs [hunk 5]
```diff
@@ -16,6 +16,8 @@ public class IOconfTemp : IOconfInput
         public static IOconfTemp NewTypeJ(string row, int lineNum) => new(row, lineNum, TypeJName, "0.0000579530", "0.0000521360");
         //CAL portNumber,delta,coldJunctionDelta portNumber,delta,...
         static readonly string DefaultCalibrations = $"CAL {string.Join(" ", Enumerable.Range(1, 10).Select(i => $"{i},0.0000412760,0.0000407300"))}";
+        
+        private readonly string _delta, _coldJunctionDelta;
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply CA_DataUploaderLib/IOconf/IOconfTemp.cs:20]: fixed

File: CA_DataUploaderLib/IOconf/IOconfRPiTemp.cs [hunk 6]
```diff
@@ -32,6 +33,7 @@ public IEnumerable<IOconfInput> GetDistributedExpandedInputConf()
                 //note there is no map entry for the IOconfRpiTemp as it not an external box, but at the moment we only expose the IOconfNode through it
                 var nodeNameClean = node.Name.Replace("-", "");
                 var map = new IOconfMap($"Map;RpiFakeBox;{Name}_{nodeNameClean}Box;{node.Name}", LineNumber);
+                map.ValidateDependencies(ioconf);
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply CA_DataUploaderLib/IOconf/IOconfRPiTemp.cs:36]: done

File: UnitTests/IOconfCurrentTests.cs [hunk 7]
```diff
@@ -17,14 +18,16 @@ public class IOconfCurrentTests
         public void BoardWithoutCalibrationDoesNotGetUpdated()
         {
             // Arrange
-            IOconfMap? mapLine = null;
-            using var _ = TestableIOconfFile.Override(new() { GetMap = () => new[] { mapLine = new IOconfMap($"Map; {portName}; {boxName}", 0) } });
+            var mapLine = new IOconfMap($"Map; {portName}; {boxName}", 0);
+            var ioConfMock = GenereateIOconfMock(mapLine);
+            mapLine.ValidateDependencies(ioConfMock);
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply UnitTests/IOconfCurrentTests.cs:23]: fixed

File: UnitTests/IOconfCurrentTests.cs [hunk 8]
```diff
@@ -17,14 +18,16 @@ public class IOconfCurrentTests
         public void BoardWithoutCalibrationDoesNotGetUpdated()
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply UnitTests/IOconfCurrentTests.cs:18]: fixed

File: CA_DataUploaderLib/IOconf/IOconfFile.cs [hunk 9]
```diff
@@ -6,31 +6,34 @@
...
             Table.Clear();
             var (rawLines, entries) = IOconfFileLoader.Load();
```
[the remark this answers is not in the document]
Comment (LarsWithCA) [reply CA_DataUploaderLib/IOconf/IOconfFile.cs:29]: done

**Review by freddyrios — APPROVED** [review]

**Pull request merged**